File: archive/2026-04-05/ch07_backtracking/t5_combination_sum_2.py
"""LeetCode 40. Combination Sum II
https://leetcode.cn/problems/combination-sum-ii/
"""

from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        path: List[int] = []
        ans: List[List[int]] = []
        used: List[bool] = [False] * len(candidates)

        def backtracking(nums: List[int], target: int, sum: int, start_idx: int) -> None:
            if sum == target:
                ans.append(path[:])
                return

            for i in range(start_idx, len(nums)):
                if sum > target:
                    break
                if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                    continue

                path.append(nums[i])
                used[i] = True
                backtracking(nums, target, sum + nums[i], i + 1)
                path.pop()
                used[i] = False

        if candidates:
            candidates.sort()
            backtracking(candidates, target, 0, 0)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = sorted(sorted(x) for x in solution.combinationSum2([10, 1, 2, 7, 6, 1, 5], 8))
    expected = sorted(sorted(x) for x in [[1, 1, 6], [1, 2, 5], [1, 7], [2, 6]])
    assert result == expected

    solution = Solution()
    result = sorted(sorted(x) for x in solution.combinationSum2([2, 5, 2, 1, 2], 5))
    expected = sorted(sorted(x) for x in [[1, 2, 2], [5]])
    assert result == expected

    print("All tests passed.")

File: archive/2024-08-01/backtracking/T5_combinationSum2.java
package org.rereview2024summer.backtracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class T5_combinationSum2 {
    List<Integer> path = new ArrayList<>();
    List<List<Integer>> res = new ArrayList<>();
    boolean[] used;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        used = new boolean[candidates.length];
        Arrays.sort(candidates);
        backtracking(candidates, target, 0, 0);
        return res;
    }
    public void backtracking(int[] nums, int target, int startIdx, int sum) {
        if (sum == target) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = startIdx; i < nums.length && sum + nums[i] <= target; i++) {
            if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1])
                continue;
            path.add(nums[i]);
            used[i] = true;
            backtracking(nums, target, i + 1, sum + nums[i]);
            used[i] = false;
            path.remove(path.size() - 1);
        }
    }
}
